File: php/src/Resources/TranscriptionResource.php
<?php

declare(strict_types=1);

namespace JewelMusic\Resources;

/**
 * Transcription resource for AI-powered music transcription and lyrics processing
 */
class TranscriptionResource extends BaseResource
{
    /**
     * Create a new transcription from track ID or uploaded file
     *
     * @param string|null $trackId Existing track ID (optional)
     * @param string|resource|null $file File path or file resource (optional)
     * @param string|null $filename Original filename (required if file provided)
     * @param array $options Transcription options
     *                      - languages: Target languages for transcription
     *                      - includeTimestamps: Include word-level timestamps
     *                      - wordLevelTimestamps: Enable word-level timing
     *                      - speakerDiarization: Enable speaker identification
     *                      - model: Transcription model to use
     *                      - maxSpeakers: Maximum number of speakers
     * @return array Transcription data
     */
    public function create($trackId = null, $file = null, ?string $filename = null, array $options = []): array
    {
        if ($trackId) {
            // Create from existing track
            $data = array_merge(['trackId' => $trackId], $this->filterNullValues($options));
            $response = $this->httpClient->post('/transcription/create', $data);
            return $this->extractData($response);
        }
        
        if ($file && $filename) {
            // Create from uploaded file
            $metadata = [];
            foreach ($options as $key => $value) {
                if (is_array($value)) {
                    $metadata[$key] = implode(',', $value);
                } elseif (is_bool($value)) {
                    $metadata[$key] = $value ? 'true' : 'false';
                } else {
                    $metadata[$key] = (string)$value;
                }
            }
            
            $response = $this->httpClient->uploadFile('/transcription/create', $file, $filename, $metadata);
            return $this->extractData($response);
        }
        
        throw new \InvalidArgumentException('Either trackId or file with filename must be provided');
    }

    /**
     * Get transcription by ID
     *
     * @param string $transcriptionId Transcription ID
     * @return array Transcription data
     */
    public function get(string $transcriptionId): array
    {
        $response = $this->httpClient->get("/transcription/{$transcriptionId}");
        return $this->extractData($response);
    }

    /**
     * Get transcription status and progress
     *
     * @param string $transcriptionId Transcription ID
     * @return array Status information
     */
    public function getStatus(string $transcriptionId): array
    {
        $response = $this->httpClient->get("/transcription/{$transcriptionId}/status");
        return $this->extractData($response);
    }

    /**
     * Download transcription in specified format
     *
     * @param string $transcriptionId Transcription ID
     * @param string $format Download format (srt, vtt, txt, json, lrc)
     * @return array Download data or URL
     */
    public function download(string $transcriptionId, string $format = 'json'): array
    {
        $params = ['format' => $format];
        $response = $this->httpClient->get("/transcription/{$transcriptionId}/download", $params);
        return $this->extractData($response);
    }

    /**
     * Translate lyrics to target languages
     *
     * @param string $transcriptionId Transcription ID
     * @param array $targetLanguages Target languages for translation
     * @param array $options Translation options
     *                      - preserveRhyme: Maintain rhyme scheme
     *                      - preserveMeter: Maintain musical meter
     *                      - adaptCulturally: Adapt for cultural context
     * @return array Translation results
     */
    public function translateLyrics(string $transcriptionId, array $targetLanguages, array $options = []): array
    {
        $this->validateRequired([
            'transcriptionId' => $transcriptionId,
            'targetLanguages' => $targetLanguages
        ], ['transcriptionId', 'targetLanguages']);
        
        $data = array_merge(
            ['targetLanguages' => $targetLanguages],
            $this->filterNullValues($options)
        );
        
        $response = $this->httpClient->post("/transcription/{$transcriptionId}/translate", $data);
        return $this->extractData($response);
    }

    /**
     * Synchronize lyrics with audio file
     *
     * @param string $transcriptionId Transcription ID
     * @param string|resource $audioFile Audio file for synchronization
     * @param string $filename Audio filename
     * @return array Synchronization result
     */
    public function syncLyrics(string $transcriptionId, $audioFile, string $filename): array
    {
        $this->validateRequired([
            'transcriptionId' => $transcriptionId,
            'audioFile' => $audioFile,
            'filename' => $filename
        ], ['transcriptionId', 'audioFile', 'filename']);
        
        $response = $this->httpClient->uploadFile("/transcription/{$transcriptionId}/sync", $audioFile, $filename);
        return $this->extractData($response);
    }

    /**
     * Enhance lyrics with AI assistance
     *
     * @param string $lyrics Raw lyrics text
     * @param array $options Enhancement options
     *                      - improveMeter: Improve rhythmic meter
     *                      - enhanceRhyming: Enhance rhyme scheme
     *                      - adjustTone: Adjust emotional tone
     *                      - targetLanguage: Target language for enhancement
     *                      - preserveStyle: Preserve original style
     * @return array Enhanced lyrics
     */
    public function enhanceLyrics(string $lyrics, array $options = []): array
    {
        $this->validateRequired(['lyrics' => $lyrics], ['lyrics']);
        
        $data = array_merge(['lyrics' => $lyrics], $this->filterNullValues($options));
        $response = $this->httpClient->post('/transcription/enhance-lyrics', $data);
        return $this->extractData($response);
    }

    /**
     * Analyze rhyme scheme in lyrics
     *
     * @param string $lyrics Lyrics text to analyze
     * @return array Rhyme scheme analysis
     */
    public function checkRhymeScheme(string $lyrics): array
    {
        $this->validateRequired(['lyrics' => $lyrics], ['lyrics']);
        
        $data = ['lyrics' => $lyrics];
        $response = $this->httpClient->post('/transcription/check-rhyme-scheme', $data);
        return $this->extractData($response);
    }

    /**
     * Analyze sentiment and emotion in lyrics
     *
     * @param string $lyrics Lyrics text to analyze
     * @return array Sentiment analysis results
     */
    public function analyzeSentiment(string $lyrics): array
    {
        $this->validateRequired(['lyrics' => $lyrics], ['lyrics']);
        
        $data = ['lyrics' => $lyrics];
        $response = $this->httpClient->post('/transcription/analyze-sentiment', $data);
        return $this->extractData($response);
    }

    /**
     * Check language quality and grammar
     *
     * @param string $lyrics Lyrics text to check
     * @param string $language Language code for checking
     * @return array Language quality analysis
     */
    public function checkLanguageQuality(string $lyrics, string $language): array
    {
        $this->validateRequired([
            'lyrics' => $lyrics,
            'language' => $language
        ], ['lyrics', 'language']);
        
        $data = [
            'lyrics' => $lyrics,
            'language' => $language
        ];
        
        $response = $this->httpClient->post('/transcription/check-language-quality', $data);
        return $this->extractData($response);
    }

    /**
     * Generate alternative lyrics suggestions
     *
     * @param string $lyrics Original lyrics
     * @param array $options Generation options
     *                      - style: Musical style context
     *                      - mood: Target mood
     *                      - themes: Thematic elements to include
     *                      - constraints: Creative constraints
     * @return array Alternative lyrics suggestions
     */
    public function generateAlternatives(string $lyrics, array $options = []): array
    {
        $this->validateRequired(['lyrics' => $lyrics], ['lyrics']);
        
        $data = array_merge(['lyrics' => $lyrics], $this->filterNullValues($options));
        $response = $this->httpClient->post('/transcription/generate-alternatives', $data);
        return $this->extractData($response);
    }

    /**
     * Extract vocal parts from multi-track audio
     *
     * @param string|resource $audioFile Audio file containing multiple tracks
     * @param string $filename Audio filename
     * @param array $options Vocal extraction options
     *                      - isolationMethod: Method for vocal isolation
     *                      - enhanceVocals: Enhance extracted vocals
     *                      - removeBackingVocals: Remove backing vocals
     * @return array Vocal extraction results
     */
    public function extractVocals($audioFile, string $filename, array $options = []): array
    {
        $metadata = [];
        foreach ($options as $key => $value) {
            if (is_bool($value)) {
                $metadata[$key] = $value ? 'true' : 'false';
            } else {
                $metadata[$key] = (string)$value;
            }
        }
        
        $response = $this->httpClient->uploadFile('/transcription/extract-vocals', $audioFile, $filename, $metadata);
        return $this->extractData($response);
    }

    /**
     * Convert lyrics between different formats
     *
     * @param string $transcriptionId Transcription ID
     * @param string $targetFormat Target format (srt, vtt, lrc, txt, json)
     * @param array $options Conversion options
     *                      - includeTimestamps: Include timing information
     *                      - wordLevel: Word-level granularity
     * @return array Converted lyrics
     */
    public function convertFormat(string $transcriptionId, string $targetFormat, array $options = []): array
    {
        $data = array_merge(
            ['targetFormat' => $targetFormat],
            $this->filterNullValues($options)
        );
        
        $response = $this->httpClient->post("/transcription/{$transcriptionId}/convert", $data);
        return $this->extractData($response);
    }

    /**
     * Batch process multiple transcriptions
     *
     * @param array $transcriptionIds Array of transcription IDs
     * @param array $operations Operations to perform
     * @param array $options Batch processing options
     * @return array Batch processing results
     */
    public function batchProcess(array $transcriptionIds, array $operations, array $options = []): array
    {
        $data = array_merge([
            'transcriptionIds' => $transcriptionIds,
            'operations' => $operations
        ], $this->filterNullValues($options));
        
        $response = $this->httpClient->post('/transcription/batch-process', $data);
        return $this->extractData($response);
    }

    /**
     * List transcriptions with filtering and pagination
     *
     * @param int $page Page number
     * @param int $perPage Items per page
     * @param array $filters Transcription filters
     *                      - status: Processing status
     *                      - language: Language filter
     *                      - dateRange: Creation date range
     * @return array Transcriptions list
     */
    public function list(int $page = 1, int $perPage = 20, array $filters = []): array
    {
        $params = $this->buildParams([
            'page' => (string)$page,
            'perPage' => (string)$perPage
        ], $filters);
        
        $response = $this->httpClient->get('/transcription', $params);
        return $this->extractData($response);
    }
}

File: ruby/lib/jewelmusic/resources/transcription_resource.rb
# frozen_string_literal: true

module JewelMusic
  module Resources
    ##
    # Transcription resource for AI-powered music transcription and lyrics processing
    #
    class TranscriptionResource < BaseResource
      ##
      # Create a new transcription from track ID or uploaded file
      #
      # @param track_id [String, nil] Existing track ID (optional)
      # @param file [String, File, IO, nil] File path or file object (optional)
      # @param filename [String, nil] Original filename (required if file provided)
      # @param options [Hash] Transcription options
      # @option options [Array<String>] :languages Target languages for transcription
      # @option options [Boolean] :include_timestamps Include word-level timestamps
      # @option options [Boolean] :word_level_timestamps Enable word-level timing
      # @option options [Boolean] :speaker_diarization Enable speaker identification
      # @option options [String] :model Transcription model to use
      # @option options [Integer] :max_speakers Maximum number of speakers
      # @return [Hash] Transcription data
      #
      def create(track_id: nil, file: nil, filename: nil, **options)
        if track_id
          # Create from existing track
          data = { track_id: track_id }.merge(filter_nil_values(options))
          response = http_client.post('/transcription/create', data)
          return extract_data(response)
        end
        
        if file && filename
          # Create from uploaded file
          prepared_file = prepare_file(file)
          metadata = prepare_upload_metadata(options)
          
          response = http_client.upload_file('/transcription/create', prepared_file, filename, metadata)
          return extract_data(response)
        end
        
        raise ArgumentError, 'Either track_id or file with filename must be provided'
      end

      ##
      # Get transcription by ID
      #
      # @param transcription_id [String] Transcription ID
      # @return [Hash] Transcription data
      #
      def get(transcription_id)
        validate_required({ transcription_id: transcription_id }, [:transcription_id])
        response = http_client.get("/transcription/#{transcription_id}")
        extract_data(response)
      end

      ##
      # Get transcription status and progress
      #
      # @param transcription_id [String] Transcription ID
      # @return [Hash] Status information
      #
      def get_status(transcription_id)
        validate_required({ transcription_id: transcription_id }, [:transcription_id])
        response = http_client.get("/transcription/#{transcription_id}/status")
        extract_data(response)
      end

      ##
      # Download transcription in specified format
      #
      # @param transcription_id [String] Transcription ID
      # @param format [String] Download format (srt, vtt, txt, json, lrc)
      # @return [Hash] Download data or URL
      #
      def download(transcription_id, format = 'json')
        validate_required({ transcription_id: transcription_id }, [:transcription_id])
        params = { format: format }
        response = http_client.get("/transcription/#{transcription_id}/download", params)
        extract_data(response)
      end

      ##
      # Translate lyrics to target languages
      #
      # @param transcription_id [String] Transcription ID
      # @param target_languages [Array<String>] Target languages for translation
      # @param options [Hash] Translation options
      # @option options [Boolean] :preserve_rhyme Maintain rhyme scheme
      # @option options [Boolean] :preserve_meter Maintain musical meter
      # @option options [Boolean] :adapt_culturally Adapt for cultural context
      # @return [Hash] Translation results
      #
      def translate_lyrics(transcription_id, target_languages, options = {})
        validate_required({
          transcription_id: transcription_id,
          target_languages: target_languages
        }, [:transcription_id, :target_languages])
        
        data = {
          target_languages: target_languages
        }.merge(filter_nil_values(options))
        
        response = http_client.post("/transcription/#{transcription_id}/translate", data)
        extract_data(response)
      end

      ##
      # Synchronize lyrics with audio file
      #
      # @param transcription_id [String] Transcription ID
      # @param audio_file [String, File, IO] Audio file for synchronization
      # @param filename [String] Audio filename
      # @return [Hash] Synchronization result
      #
      def sync_lyrics(transcription_id, audio_file, filename)
        validate_required({
          transcription_id: transcription_id,
          audio_file: audio_file,
          filename: filename
        }, [:transcription_id, :audio_file, :filename])
        
        prepared_file = prepare_file(audio_file)
        response = http_client.upload_file("/transcription/#{transcription_id}/sync", prepared_file, filename)
        extract_data(response)
      end

      ##
      # Enhance lyrics with AI assistance
      #
      # @param lyrics [String] Raw lyrics text
      # @param options [Hash] Enhancement options
      # @option options [Boolean] :improve_meter Improve rhythmic meter
      # @option options [Boolean] :enhance_rhyming Enhance rhyme scheme
      # @option options [String] :adjust_tone Adjust emotional tone
      # @option options [String] :target_language Target language for enhancement
      # @option options [Boolean] :preserve_style Preserve original style
      # @return [Hash] Enhanced lyrics
      #
      def enhance_lyrics(lyrics, options = {})
        validate_required({ lyrics: lyrics }, [:lyrics])
        
        data = { lyrics: lyrics }.merge(filter_nil_values(options))
        response = http_client.post('/transcription/enhance-lyrics', data)
        extract_data(response)
      end

      ##
      # Analyze rhyme scheme in lyrics
      #
      # @param lyrics [String] Lyrics text to analyze
      # @return [Hash] Rhyme scheme analysis
      #
      def check_rhyme_scheme(lyrics)
        validate_required({ lyrics: lyrics }, [:lyrics])
        
        data = { lyrics: lyrics }
        response = http_client.post('/transcription/check-rhyme-scheme', data)
        extract_data(response)
      end

      ##
      # Analyze sentiment and emotion in lyrics
      #
      # @param lyrics [String] Lyrics text to analyze
      # @return [Hash] Sentiment analysis results
      #
      def analyze_sentiment(lyrics)
        validate_required({ lyrics: lyrics }, [:lyrics])
        
        data = { lyrics: lyrics }
        response = http_client.post('/transcription/analyze-sentiment', data)
        extract_data(response)
      end

      ##
      # Check language quality and grammar
      #
      # @param lyrics [String] Lyrics text to check
      # @param language [String] Language code for checking
      # @return [Hash] Language quality analysis
      #
      def check_language_quality(lyrics, language)
        validate_required({ lyrics: lyrics, language: language }, [:lyrics, :language])
        
        data = {
          lyrics: lyrics,
          language: language
        }
        
        response = http_client.post('/transcription/check-language-quality', data)
        extract_data(response)
      end

      ##
      # Generate alternative lyrics suggestions
      #
      # @param lyrics [String] Original lyrics
      # @param options [Hash] Generation options
      # @option options [String] :style Musical style context
      # @option options [String] :mood Target mood
      # @option options [Array<String>] :themes Thematic elements to include
      # @option options [Hash] :constraints Creative constraints
      # @return [Hash] Alternative lyrics suggestions
      #
      def generate_alternatives(lyrics, options = {})
        validate_required({ lyrics: lyrics }, [:lyrics])
        
        data = { lyrics: lyrics }.merge(filter_nil_values(options))
        response = http_client.post('/transcription/generate-alternatives', data)
        extract_data(response)
      end

      ##
      # Extract vocal parts from multi-track audio
      #
      # @param audio_file [String, File, IO] Audio file containing multiple tracks
      # @param filename [String] Audio filename
      # @param options [Hash] Vocal extraction options
      # @option options [String] :isolation_method Method for vocal isolation
      # @option options [Boolean] :enhance_vocals Enhance extracted vocals
      # @option options [Boolean] :remove_backing_vocals Remove backing vocals
      # @return [Hash] Vocal extraction results
      #
      def extract_vocals(audio_file, filename, options = {})
        validate_required({ audio_file: audio_file, filename: filename }, [:audio_file, :filename])
        
        prepared_file = prepare_file(audio_file)
        metadata = prepare_upload_metadata(options)
        
        response = http_client.upload_file('/transcription/extract-vocals', prepared_file, filename, metadata)
        extract_data(response)
      end

      ##
      # Convert lyrics between different formats
      #
      # @param transcription_id [String] Transcription ID
      # @param target_format [String] Target format (srt, vtt, lrc, txt, json)
      # @param options [Hash] Conversion options
      # @option options [Boolean] :include_timestamps Include timing information
      # @option options [Boolean] :word_level Word-level granularity
      # @return [Hash] Converted lyrics
      #
      def convert_format(transcription_id, target_format, options = {})
        validate_required({ transcription_id: transcription_id, target_format: target_format }, [:transcription_id, :target_format])
        
        data = {
          target_format: target_format
        }.merge(filter_nil_values(options))
        
        response = http_client.post("/transcription/#{transcription_id}/convert", data)
        extract_data(response)
      end

      ##
      # Batch process multiple transcriptions
      #
      # @param transcription_ids [Array<String>] Array of transcription IDs
      # @param operations [Array<String>] Operations to perform
      # @param options [Hash] Batch processing options
      # @return [Hash] Batch processing results
      #
      def batch_process(transcription_ids, operations, options = {})
        validate_required({ transcription_ids: transcription_ids, operations: operations }, [:transcription_ids, :operations])
        
        data = {
          transcription_ids: transcription_ids,
          operations: operations
        }.merge(filter_nil_values(options))
        
        response = http_client.post('/transcription/batch-process', data)
        extract_data(response)
      end

      ##
      # List transcriptions with filtering and pagination
      #
      # @param page [Integer] Page number
      # @param per_page [Integer] Items per page
      # @param filters [Hash] Transcription filters
      # @option filters [String] :status Processing status
      # @option filters [String] :language Language filter
      # @option filters [String] :date_range Creation date range
      # @return [Hash] Transcriptions list
      #
      def list(page = 1, per_page = 20, filters = {})
        params = build_params({
          page: page.to_s,
          per_page: per_page.to_s
        }, filters)
        
        response = http_client.get('/transcription', params)
        extract_data(response)
      end
    end
  end
end